File: 17-decorators_vs_wrappers.rb
# Ruby 冇 Python 式 decorator，但可以用 define_singleton_method 包住 method。
def log_call(method_name)
  original_method = method(method_name)

  define_singleton_method(method_name) do |*args|
    puts "Calling #{method_name}"
    result = original_method.call(*args)
    puts "Finished #{method_name}"

    result
  end
end

def total(items)
  items.sum
end

log_call(:total)

puts total([10, 20, 30])

puts '---'

# block/yield 好適合寫 retry 呢類 wrapper 流程。
class Retry
  def initialize(attempts:)
    @attempts = attempts
  end

  def run
    last_error = nil

    (1..@attempts).each do |attempt|
      return yield
    rescue RuntimeError => e
      last_error = e
      puts "Attempt #{attempt} failed"
    end

    raise last_error
  end
end

counter = { count: 0 }

sometimes_fails = lambda do
  counter[:count] += 1

  raise 'Not ready' if counter[:count] < 2

  'OK'
end

puts Retry.new(attempts: 3).run(&sometimes_fails)

File: 17-decorators_vs_wrappers.py
def log_call(function):
    def wrapper(*args, **kwargs):
        print(f"Calling {function.__name__}")
        result = function(*args, **kwargs)
        print(f"Finished {function.__name__}")

        return result

    return wrapper


@log_call
def total(items):
    return sum(items)


print(total([10, 20, 30]))

print("---")


class Retry:
    def __init__(self, attempts):
        self.attempts = attempts

    def run(self, function):
        last_error = None

        for attempt in range(1, self.attempts + 1):
            try:
                return function()
            except RuntimeError as error:
                last_error = error
                print(f"Attempt {attempt} failed")

        raise last_error


counter = {"count": 0}


def sometimes_fails():
    counter["count"] += 1

    if counter["count"] < 2:
        raise RuntimeError("Not ready")

    return "OK"


print(Retry(attempts=3).run(sometimes_fails))
